### DISCUSSION TITLE
Writing to File Attachment or other ways to persist data?

### DISCUSSION THREAD

**Discussion by aaronkyle** [observablehq/framework#1723] [Q&A]
Does Observable Framework make it possible to write data back into a local file that is saved in a Project folder and referenced as a file attachment – such that you could change the locally-stored data object with the notebook and the updated values would be used next time you load the application?

My assumption is no: the browser still runs in a sandboxed environment that prevents access to the local file system.

But how about with the File System Access API? … Looks like it might be possible, but only with Edge? [ref: [Window: showSaveFilePicker() method - Web APIs | MDN](https://developer.mozilla.org/en-US/docs/Web/API/Window/showSaveFilePicker)]

I’d be interested if anyone could share an example of using Framework to persist changes to data in a relatively simple way.

**Comment by Fil** [comment]
I am doing this right now with an `Input.textarea({value: JSON.stringify(data)})`; then I have to copy from the textarea and go to my terminal and type `pbpaste > src/data/reference.json`. A download button could serve a similar purpose (#869).

**Comment by aaronkyle** [comment]
Thank you, @Fil .  Reflecting on the nature of Framework, what I am thinking about probably isn't tenable--even if I can write data back to a local data object, it might not work to be able to reference that object without rebuilding the site?  The broader context of what I am exploring is this:  If I have a multi-page site, each page can reference a central data object, but changes to inputs on one page don't appear on my other pages.  To get around this, I have been making an 'ephemeral' data object in the browser memory, then calling on that object in other pages instead of calling on the original. (e.g. https://observablehq.com/@aaronkyle/ephemeral-data-changes) This sorta works (and allows quasi persistent changes), but at some point I'd like to be able to write back my updates more permanently without manual intervention -- like after deploying an app to cloud.

**Reply by Fil** [reply]
I think what you're describing is a database backend (or “KV store”). You will need to set up a server or use a service provider.
